compiler-core/src/language_server/engine.rs

```diff
@@ -542,3 +659,142 @@ fn hover_for_expression(expression: &TypedExpr, line_numbers: LineNumbers) -> Ho
         range: Some(src_span_to_lsp_range(expression.location(), &line_numbers)),
     }
 }
+
+fn is_action_attached_to_unused_diagnostic(params: &lsp::CodeActionParams) -> bool {
+    params
+        .context
+        .diagnostics
+        .iter()
+        .any(|diag| diag.message.starts_with("Unused"))
+}
+
+fn store_unused_warning_range(
+    cache: &mut HashMap<Utf8PathBuf, Vec<lsp_types::Range>>,
+    path: &Utf8PathBuf,
+    range: lsp_types::Range,
+) {
+    match cache.get_mut(path) {
+        None => {
+            let _ = cache.insert(path.clone(), vec![range]);
+        }
+        Some(vec) => vec.push(range),
+    }
+}
+
+fn import_token_location(before: &str) -> Option<usize> {
+    let pos = before.rfind("import")?;
+    // Consume the previous line return if it exists.
+    Some(if pos > 0 { pos - 1 } else { pos })
+}
+
+fn mk_range(line_numbers: &LineNumbers, start: usize, end: usize) -> lsp_types::Range {
+    let location = SrcSpan {
+        start: start as u32,
+        end: end as u32,
+    };
+    src_span_to_lsp_range(location, line_numbers)
+}
+
+// This function handle UnusedImportedModule warning.
+// It finds the full range of the import statement,
+// because the warning only contains the location of the module name.
+fn handle_unused_imported_module(
+    line_numbers: &LineNumbers,
+    code: &SmolStr,
+    location: &SrcSpan,
+) -> Option<lsp_types::Range> {
+    let (before, after) = code.split_at(location.start as usize);
+    // Find the previous import keyword.
+    let start = import_token_location(before)?;
+    // Find the next line return.
+    let end = location.start as usize + (after.find('\n')?);
+    // The src span of the full import statement.
+    Some(mk_range(line_numbers, start, end))
+}
+
+#[derive(Debug)]
+struct QualifiedUnusedInfo {
+    unqualified_count: usize,
+    has_name: bool,
+    location: SrcSpan,
+    unused: Vec<SrcSpan>,
+}
+fn handle_unused_imported_module_qualified(
```

Comment:
As explained below this is to include the surrounding comma so that unqualified name can be safely removed. I moved this logic to the parser in 8d8af27, though I'm not sure that's better.